rust/parquet_derive/src/parquet_field.rs

```diff
@@ -0,0 +1,942 @@
+
+            if is_vec {
+                Type::Vec(Box::new(Type::from_type(f, &generic_type)))
+            } else {
+                Type::Option(Box::new(Type::from_type(f, &generic_type)))
+            }
+        } else {
+            Type::TypePath(syn::Type::Path(p.clone()))
+        }
+    }
+
+    fn from_type_reference(f: &syn::Field, tr: &syn::TypeReference) -> Self {
+        let lifetime = tr.lifetime.clone();
+        let inner_type = Type::from_type(f, tr.elem.as_ref());
+        Type::Reference(lifetime, Box::new(inner_type))
+    }
+
+    fn from_type_array(f: &syn::Field, ta: &syn::TypeArray) -> Self {
+        let inner_type = Type::from_type(f, ta.elem.as_ref());
+        Type::Array(Box::new(inner_type))
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use syn::{self, Data, DataStruct, DeriveInput};
+
+    fn extract_fields(input: proc_macro2::TokenStream) -> Vec<syn::Field> {
+        let input: DeriveInput = syn::parse2(input).unwrap();
+
+        let fields = match input.data {
+            Data::Struct(DataStruct { fields, .. }) => fields,
+            _ => panic!("input must be a struct"),
+        };
+
+        fields.iter().map(|x| x.to_owned()).collect()
+    }
+
+    #[test]
+    fn test_generating_a_simple_writer_snippet() {
+        let snippet: proc_macro2::TokenStream = quote! {
+          struct ABoringStruct {
+            counter: usize,
+          }
+        };
+
+        let fields = extract_fields(snippet);
+        let counter = Field::from(&fields[0]);
+
+        let snippet = counter.writer_snippet().to_string();
+        assert_eq!(snippet,
+                   (quote!{
+                        {
+                            let vals : Vec < _ > = records . iter ( ) . map ( | x | x . counter ) . collect ( );
+
+                            if let parquet::column::writer::ColumnWriter::Int64ColumnWriter ( ref mut typed ) = column_writer {
+                                typed . write_batch ( & vals [ .. ] , None , None ) . unwrap ( );
+                            }  else {
+                                panic!("schema and struct disagree on type for {}" , stringify!{ counter } )
+                            }
+                        }
+                   }).to_string()
+        )
+    }
+
+    #[test]
+    fn test_optional_to_writer_snippet() {
+        let struct_def: proc_macro2::TokenStream = quote! {
+          struct StringBorrower<'a> {
+            optional_str: Option<&'a str>,
+            optional_string: &Option<String>,
+            optional_dumb_int: &Option<&i32>,
+          }
+        };
+
+        let fields = extract_fields(struct_def);
+
+        let optional = Field::from(&fields[0]);
+        let snippet = optional.writer_snippet();
+        assert_eq!(snippet.to_string(),
+          (quote! {
+          {
+                let definition_levels : Vec < i16 > = self . iter ( ) . map ( | x | if x . optional_str . is_some ( ) { 1 } else { 0 } ) . collect ( ) ;
```

Comment:
Why there are spaces between the words? let's limit the line width to 90 characters.

Reply:
The spaces between words are how the AST is printed out from the `to_string()` method from the syn library. It's the only way I know of to compare derived code to expected code.